File: src/main/swift/g0101_0200/s0198_house_robber/Solution.swift
// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Array #Dynamic_Programming
// #Algorithm_I_Day_12_Dynamic_Programming #Dynamic_Programming_I_Day_3
// #Level_2_Day_12_Dynamic_Programming #Udemy_Dynamic_Programming #Big_O_Time_O(n)_Space_O(n)
// #2024_06_28_Time_0_ms_(100.00%)_Space_15.6_MB_(25.32%)

class Solution {
    func rob(_ nums: [Int]) -> Int {
        guard nums.count > 0 else {
            return 0
        }
        if nums.count == 1 {
            return nums[0]
        }
        if nums.count == 2 {
            return max(nums[0], nums[1])
        }
        
        var profit = [Int](repeating: 0, count: nums.count)
        profit[0] = nums[0]
        profit[1] = max(nums[0], nums[1])
        
        for i in 2..<nums.count {
            profit[i] = max(profit[i - 1], nums[i] + profit[i - 2])
        }
        return profit[nums.count - 1]
    }
}

File: src/main/rust/g0101_0200/s0198_house_robber/Solution.rs
// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Array #Dynamic_Programming
// #Algorithm_I_Day_12_Dynamic_Programming #Dynamic_Programming_I_Day_3
// #Level_2_Day_12_Dynamic_Programming #Udemy_Dynamic_Programming #Big_O_Time_O(n)_Space_O(n)
// #2024_09_10_Time_0_ms_(100.00%)_Space_2.1_MB_(31.95%)

impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        if nums.is_empty() {
            return 0;
        }
        if nums.len() == 1 {
            return nums[0];
        }
        if nums.len() == 2 {
            return std::cmp::max(nums[0], nums[1]);
        }

        let mut profit = vec![0; nums.len()];
        profit[0] = nums[0];
        profit[1] = std::cmp::max(nums[0], nums[1]);

        for i in 2..nums.len() {
            profit[i] = std::cmp::max(profit[i - 1], nums[i] + profit[i - 2]);
        }

        profit[nums.len() - 1]
    }
}
